### PR TITLE
Unconditional recursion

### LINKED ISSUE #412: new rule: unconditional-recursion
Unconditional recursive calls will produce infinite recursion, thus program stack overflow.
Even if it seems that the frequency of such errors is low, autocompletion functionality of IDEs facilitates them. A typical case is:

```go
type myType struct {
  field int
}

func (mt myType) Field() int {
  return mt.Field()          // /!\ recursive call in place of accessing mt.field
} 
```

### PR DESCRIPTION
**Pull request by chavacava**
Closes #412

### CODE REVIEW

File: rule/unconditional-recursion.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+// Apply applies the rule to given file.
+func (r *UnconditionalRecursionRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	var failures []lint.Failure
+
+	onFailure := func(failure lint.Failure) {
+		failures = append(failures, failure)
+	}
+
+	// err := file.Pkg.TypeCheck()
```
Comment (mgechev) [inline rule/unconditional-recursion.go:20]: We can remove the type checking here?

File: rule/unconditional-recursion.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		default:
+			return w
+		}
+
+		if w.currentFunc != nil && // not in a func body
+			!w.currentFunc.seenConditionalExit && // there is a conditional exit in the function
+			w.currentFunc.funcDesc.equal(&funcDesc{selector, funcId}) {
+			w.onFailure(lint.Failure{
+				Category:   "logic",
+				Confidence: 1,
```
Comment (mgechev) [inline rule/unconditional-recursion.go:98]: Maybe we can decrease the confidence a bit? A construct with a for loop when we're draining a channel may still throw?

File: rule/unconditional-recursion.go (hunk 1 above)
Comment (chavacava) [reply rule/unconditional-recursion.go:20]: sure, type-checking is not necessary for this rule

File: rule/unconditional-recursion.go (hunk 2 above)
Comment (chavacava) [reply rule/unconditional-recursion.go:98]: The rule is all about control dependencies.
`for-range`'s bodies are analysed looking for stmt taking the control out of the function (`return`, `os.Exit`, ...) to check if subsequent statements depend upon the loop
A channel-draining loop is not different from other loop.

About the confidence: I do not see cases where we can generate a false positive (do you have examples in mind). Contrary we can pass silently recursive calls, for example:
```go
func foo() {
  for false { // or other arbitrary complex condition evaluating to false
    return
  }
  foo()
``` 
will not trigger a failure.

**Comment by mgechev** [comment]
@chavacava there are some conflicts with `master`. We can merge after that.

**Pull request merged**